### PR TITLE
[Replica Level Throttle] Add per replica rebalance type compute logic

### PR DESCRIPTION
**Pull request by dasahcc**
### Issues

- [x] My PR addresses the following Helix issues and references them in the PR description:

resolve #343 

### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:
The first step for per replica level throttling is supporting logic to determine what message is recovery/rebalance or load rebalance.
Three functions added: 1) rebalance type computation required state.
                                        2) rebalance type per message
                                        3) message sorting rules and comparators to determine which message to apply first.

For global change contexts please refer: https://github.com/apache/helix/pull/1713
### Tests

Split PRs. The branch will not able to compile or test until finishes code change.
Will have tests files and testing result in last several PRs.

- [ ] The following tests are written for this issue:

- The following is the result of the "mvn test" command on the appropriate module:

### Documentation (Optional)

(Link the GitHub wiki you added)

### Commits

- My commits all reference appropriate Apache Helix GitHub issues in their subject lines. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Code Quality

- My diff has been formatted using helix-style.xml 
(helix-style-intellij.xml if IntelliJ IDE is used)

### CODE REVIEW

File: helix-core/src/main/java/org/apache/helix/controller/stages/IntermediateStateCalcStage.java [hunk 1]
```diff
@@ -795,6 +798,54 @@ private RebalanceType getRebalanceType(ResourceControllerDataProvider cache,
     }
   }
 
+  private RebalanceType getRebalanceTypePerMessage(Map<String, Integer> requiredStates, Message message,
```
Comment (alirezazamani) [inline helix-core/src/main/java/org/apache/helix/controller/stages/IntermediateStateCalcStage.java:801]: This method has not been used. Right? Is it because it is skeleton PR?

Comment (dasahcc) [reply helix-core/src/main/java/org/apache/helix/controller/stages/IntermediateStateCalcStage.java:801]: Not used. I was trying to make each PR independent and easy to review. Then we can focus on the logic of the content itself.

File: helix-core/src/main/java/org/apache/helix/controller/stages/IntermediateStateCalcStage.java [hunk 2]
```diff
@@ -795,6 +798,55 @@ private RebalanceType getRebalanceType(ResourceControllerDataProvider cache,
...
+    Map<String, Integer> requiredStatesSnapshot = new HashMap<>(requiredStates);
+    // Looping existing current states to see whether current states fulfilled all the required states.
+    for (String state : derivedCurrentStates.values()) {
+      if (requiredStatesSnapshot.containsKey(state)) {
+        requiredStatesSnapshot.put(state, requiredStatesSnapshot.get(state) - 1);
+        if (requiredStatesSnapshot.get(state) == 0) {
+          requiredStatesSnapshot.remove(state);
+        }
+      }
+    }
```
Comment (jiajunwang) [inline helix-core/src/main/java/org/apache/helix/controller/stages/IntermediateStateCalcStage.java:812]: This calculation would be done multiple times if a partition has more than one message. Should we just pre-calculate the requiredStatesSnapshot (which is basically the gap between the derived current state and required states) and then use the snapshot? Obviously, we need to update the snapshot once a message is confirmed. But that calculation would be much faster than loop these 2 types of maps every time.

File: helix-core/src/main/java/org/apache/helix/controller/stages/IntermediateStateCalcStage.java [hunk 3]
```diff
@@ -874,7 +926,31 @@ public int compare(ResourcePriority priority1, ResourcePriority priority2) {
...
+      _statePriorityMap = statePriorityMap;
+    }
+
+    @Override
+    public int compare(Message m1, Message m2) {
+      //Compare rules:
+      //     1. Higher target state has higher priority.
+      //     2. If target state is same, range it as preference list order.
+      if (m1.getToState().equals(m2.getToState())) {
+        return _preferenceInstanceMap.get(m1.getTgtName()).compareTo(_preferenceInstanceMap.get(m2.getTgtName()));
```
Comment (jiajunwang) [inline helix-core/src/main/java/org/apache/helix/controller/stages/IntermediateStateCalcStage.java:947]: A message might target an instance that does not exist in the preference list.

Comment (jiajunwang) [reply helix-core/src/main/java/org/apache/helix/controller/stages/IntermediateStateCalcStage.java:947]: For safety and future extension, maybe it is a good idea to sort the messages by name as the last resort.

File: helix-core/src/main/java/org/apache/helix/controller/stages/IntermediateStateCalcStage.java [hunk 4]
```diff
@@ -795,6 +798,55 @@ private RebalanceType getRebalanceType(ResourceControllerDataProvider cache,
...
 
+  private RebalanceType getRebalanceTypePerMessage(Map<String, Integer> requiredStates, Message message,
+      Map<String, String> derivedCurrentStates) {
+    Map<String, Integer> requiredStatesSnapshot = new HashMap<>(requiredStates);
+    // Looping existing current states to see whether current states fulfilled all the required states.
+    for (String state : derivedCurrentStates.values()) {
+      if (requiredStatesSnapshot.containsKey(state)) {
+        requiredStatesSnapshot.put(state, requiredStatesSnapshot.get(state) - 1);
+        if (requiredStatesSnapshot.get(state) == 0) {
+          requiredStatesSnapshot.remove(state);
```
Comment (narendly) [inline helix-core/src/main/java/org/apache/helix/controller/stages/IntermediateStateCalcStage.java:809]: You can potentially save a `put()` call by
```
int requiredStateCount = requiredStatesSnapshot.get(state);
if (--requiredStateCount == 0) {
  requiredStatesSnapshot.remove(state);
} else {
  requiredStatesSnapshot.put(state, requiredStateCount);
}
```

File: helix-core/src/main/java/org/apache/helix/controller/stages/IntermediateStateCalcStage.java [hunk 5]
```diff
@@ -795,6 +798,55 @@ private RebalanceType getRebalanceType(ResourceControllerDataProvider cache,
     }
   }
 
+  private RebalanceType getRebalanceTypePerMessage(Map<String, Integer> requiredStates, Message message,
+      Map<String, String> derivedCurrentStates) {
+    Map<String, Integer> requiredStatesSnapshot = new HashMap<>(requiredStates);
+    // Looping existing current states to see whether current states fulfilled all the required states.
+    for (String state : derivedCurrentStates.values()) {
```
Comment (narendly) [inline helix-core/src/main/java/org/apache/helix/controller/stages/IntermediateStateCalcStage.java:805]: Did we agree that something like `desiredStateCounts` is a better description?

File: helix-core/src/main/java/org/apache/helix/controller/stages/IntermediateStateCalcStage.java [hunk 6]
```diff
@@ -795,6 +798,55 @@ private RebalanceType getRebalanceType(ResourceControllerDataProvider cache,
...
+      if (requiredStatesSnapshot.containsKey(state)) {
+        requiredStatesSnapshot.put(state, requiredStatesSnapshot.get(state) - 1);
+        if (requiredStatesSnapshot.get(state) == 0) {
+          requiredStatesSnapshot.remove(state);
+        }
+      }
+    }
+
+    // If the message is trying to bring the required state remaining in the map, it is recovery rebalance.
+    // Otherwise it is load rebalance.
```
Comment (narendly) [inline helix-core/src/main/java/org/apache/helix/controller/stages/IntermediateStateCalcStage.java:815]: We could maybe reword this: 

"If the message contains any "required" state changes, then it is considered recovery rebalance. Otherwise, it is load balance."

File: helix-core/src/main/java/org/apache/helix/controller/stages/IntermediateStateCalcStage.java [hunk 7]
```diff
@@ -795,6 +798,55 @@ private RebalanceType getRebalanceType(ResourceControllerDataProvider cache,
...
+
+    // Prepare required inputs: 1) Priority State List 2) required number of replica
+    IdealState idealState = resourceControllerDataProvider.getIdealState(resourceName);
+    StateModelDefinition stateModelDefinition =
+        resourceControllerDataProvider.getStateModelDef(idealState.getStateModelDefRef());
+    int requiredNumReplica = idealState.getMinActiveReplicas() == -1
+        ? idealState.getReplicaCount(preferenceList.size())
+        : idealState.getMinActiveReplicas();
+    Set<String> activeList = new HashSet<>(preferenceList);
+    activeList.retainAll(resourceControllerDataProvider.getEnabledLiveInstances());
```
Comment (narendly) [inline helix-core/src/main/java/org/apache/helix/controller/stages/IntermediateStateCalcStage.java:831]: I think we could avoid doing a deep copy since we only seem to need the size of live instances in the preference list?

File: helix-core/src/main/java/org/apache/helix/controller/stages/IntermediateStateCalcStage.java [hunk 8]
```diff
@@ -795,6 +798,55 @@ private RebalanceType getRebalanceType(ResourceControllerDataProvider cache,
...
+    // If the message is trying to bring the required state remaining in the map, it is recovery rebalance.
+    // Otherwise it is load rebalance.
+    return requiredStatesSnapshot.containsKey(message.getToState()) ? RebalanceType.RECOVERY_BALANCE
+        : RebalanceType.LOAD_BALANCE;
+  }
+
+  private Map<String, Integer> getRequiredStates(String resourceName,
+      ResourceControllerDataProvider resourceControllerDataProvider, List<String> preferenceList) {
+
+    // Prepare required inputs: 1) Priority State List 2) required number of replica
```
[the remark this answers is not in the document]
Comment (dasahcc) [reply helix-core/src/main/java/org/apache/helix/controller/stages/IntermediateStateCalcStage.java:823]: I split the PR for easy review. Finally, I will clear up the unused the method.

File: helix-core/src/main/java/org/apache/helix/controller/stages/IntermediateStateCalcStage.java [hunk 9]
```diff
@@ -795,6 +798,55 @@ private RebalanceType getRebalanceType(ResourceControllerDataProvider cache,
...
+        resourceControllerDataProvider.getStateModelDef(idealState.getStateModelDefRef());
+    int requiredNumReplica = idealState.getMinActiveReplicas() == -1
+        ? idealState.getReplicaCount(preferenceList.size())
+        : idealState.getMinActiveReplicas();
+    Set<String> activeList = new HashSet<>(preferenceList);
+    activeList.retainAll(resourceControllerDataProvider.getEnabledLiveInstances());
+
+    // For each state, check that this partition currently has the required number of that state as
+    // required by StateModelDefinition.
+    LinkedHashMap<String, Integer> expectedStateCountMap =
```
[the remark this answers is not in the document]
Comment (dasahcc) [reply helix-core/src/main/java/org/apache/helix/controller/stages/IntermediateStateCalcStage.java:835]: Good catch!

File: helix-core/src/main/java/org/apache/helix/controller/stages/IntermediateStateCalcStage.java (hunk 2 above)
Comment (dasahcc) [reply helix-core/src/main/java/org/apache/helix/controller/stages/IntermediateStateCalcStage.java:812]: No. We cannot, because the required state can be update and will be more complicated just update the required state. For example, in the state of: required state 1 LEADER, 1 FOLLOWER
1: FOLLOWER
2: OFFLINE

required state : 1 LEADER
with first message FOLLOWER -> LEADER applied.
required state : 1 FOLLOWER

So with this message apply, you need to give back the FOLLOWER. It will be complicated to determine when we should give the state back to required and when is not. I am not saying we cannot do that but just make the logic more complicated.

File: helix-core/src/main/java/org/apache/helix/controller/stages/IntermediateStateCalcStage.java (hunk 3 above)
Comment (dasahcc) [reply helix-core/src/main/java/org/apache/helix/controller/stages/IntermediateStateCalcStage.java:947]: I can make the sort check for that. But we dont need sort it by name. Because if the instance does not exist in preference list, it means the message for this instance should be downward or something. Then sort by state of target state will make it last priority.

File: helix-core/src/main/java/org/apache/helix/controller/stages/IntermediateStateCalcStage.java (hunk 7 above)
Comment (dasahcc) [reply helix-core/src/main/java/org/apache/helix/controller/stages/IntermediateStateCalcStage.java:831]: No. We need the enabled ones. If we dont do deep copy, the original data will be changed. This is something we learned from previous building IntermediateStage.

Comment (narendly) [reply helix-core/src/main/java/org/apache/helix/controller/stages/IntermediateStateCalcStage.java:839]: I think what you don't want is a concurrent modification exception here.

Let's see... `getEnabledLiveInstances()` already gives you a deep-copied set. And we only need an activeList size. Where are we getting this `preferenceList` from? Isn't that a deep-copy in and of itself? I can't seem to find its usage in this PR. I am suggesting doing something like `stream().filter().count()` without having to do a deep copy for each resource, which could be potentially expensive.

File: helix-core/src/main/java/org/apache/helix/controller/stages/IntermediateStateCalcStage.java (hunk 3 above)
Comment (jiajunwang) [reply helix-core/src/main/java/org/apache/helix/controller/stages/IntermediateStateCalcStage.java:948]: The reason for sorting by naming is more deterministic. I agree that there is no solid reason that we have to sort by the names. But it doesn't hurt and can help to avoid any randomness which might be annoying.

BTW, the first problem that I mentioned is not related to sorting. I believe an NPE will be thrown if you don't do a null check.

File: helix-core/src/main/java/org/apache/helix/controller/stages/IntermediateStateCalcStage.java [hunk 10]
```diff
@@ -795,6 +798,54 @@ private RebalanceType getRebalanceType(ResourceControllerDataProvider cache,
...
+  /**
+   * Determine the message rebalance type with message and current states.
+   * @param desiredStates         Ideally how may states we needed for guarantee the health of replica
+   * @param message               The message to be determined what is the rebalance type
+   * @param derivedCurrentStates  Derived from current states with previous messages not be throttled.
+   * @return                      Rebalance type. Recovery or load.
+   */
+  private RebalanceType getRebalanceTypePerMessage(Map<String, Integer> desiredStates, Message message,
+      Map<String, String> derivedCurrentStates) {
+    Map<String, Integer> requiredStatesSnapshot = new HashMap<>(desiredStates);
```
Comment (alirezazamani) [inline helix-core/src/main/java/org/apache/helix/controller/stages/IntermediateStateCalcStage.java:810]: Since this Map is created based on desiredState, how about we call it desiredStatesSnapshot?

**Review by alirezazamani — APPROVED** [review]
Overall looks good to me.

File: helix-core/src/main/java/org/apache/helix/controller/stages/IntermediateStateCalcStage.java (hunk 2 above)
Comment (jiajunwang) [reply helix-core/src/main/java/org/apache/helix/controller/stages/IntermediateStateCalcStage.java]: I think the time complexities of the 2 methods are very different. O(n^2): (n + n-1 + n-2...) vs, O(n). I get the point that logic would be a little bit complicated, but I would still prefer updating the state map on the fly instead repeat the same calculation.

File: helix-core/src/main/java/org/apache/helix/controller/stages/IntermediateStateCalcStage.java [hunk 11]
```diff
@@ -795,6 +798,54 @@ private RebalanceType getRebalanceType(ResourceControllerDataProvider cache,
...
+    StateModelDefinition stateModelDefinition =
+        resourceControllerDataProvider.getStateModelDef(idealState.getStateModelDefRef());
+    int requiredNumReplica = idealState.getMinActiveReplicas() == -1
+        ? idealState.getReplicaCount(preferenceList.size())
+        : idealState.getMinActiveReplicas();
+    Set<String> activeList = new HashSet<>(preferenceList);
+    activeList.retainAll(resourceControllerDataProvider.getEnabledLiveInstances());
+
+    // For each state, check that this partition currently has the required number of that state as
+    // required by StateModelDefinition.
```
Comment (alirezazamani) [inline helix-core/src/main/java/org/apache/helix/controller/stages/IntermediateStateCalcStage.java:842]: I couldn't quite understand the comment here. Can you please simplify it a little bit?

File: helix-core/src/main/java/org/apache/helix/controller/stages/IntermediateStateCalcStage.java (hunk 2 above)
Comment (dasahcc) [reply helix-core/src/main/java/org/apache/helix/controller/stages/IntermediateStateCalcStage.java]: Synced offline. This could involve more complicated logic such as return state back over the size of the what it required. Then the following computation will not be accurate. Will keep this as it and see whether it could cause more delay in computation.

File: helix-core/src/main/java/org/apache/helix/controller/stages/IntermediateStateCalcStage.java (hunk 7 above)
Comment (dasahcc) [reply helix-core/src/main/java/org/apache/helix/controller/stages/IntermediateStateCalcStage.java:839]: Yes. It will not be modified by other stages. ConcurrentModification should not happen.

**Comment by dasahcc** [comment]
This PR is ready to be merged.
